File: src/zig/subsystems/ffi_layout.zig
//! Portable kernels of the FFI type system: the machine-type and
//! calling-convention name tables, and the struct layout machine that assigns
//! every field its offset.
//!
//! This was the first increment inside `ffi.c`, and it deliberately takes none
//! of the calling machinery — no register classification, no argument
//! marshalling, no trampolines. What is here is arithmetic over sizes,
//! alignments, and bytes. It holds no Janet value, allocates nothing, and
//! cannot fail: an unknown name is *reported* rather than raised, because the
//! panic that names the offending Janet value belongs to `ffi_types.zig`,
//! which has the value.
//!
//! Neither `Type` nor `Struct` crosses into this file. They carry a pointer
//! into a garbage-collected abstract, and keeping the layout machine away from
//! that pointer keeps it a decision about scalars; `ffi_types.zig` dispatches
//! on the type and passes the scalars that result.
//!
//! ## Reached by import, not by symbol
//!
//! Until Phase 11 Part 16 every function here was an `export fn` that
//! `ffi_types.zig` called through a hand-written `extern fn` declaration — the
//! shape the C `ffi.c` needed, kept after both ends had become Zig. Nothing
//! checked the declaration against the definition, and `Layout` was written
//! out twice, once at each end of a symbol nothing but Zig ever called. The
//! contract in `test/ffi_layout.zig` is what removed the last reason for the
//! symbols to exist: it is compiled into the runtime and reaches these by
//! import as well, so the six exports and the second copy of `Layout` are
//! gone.
//!
//! The primitive and calling-convention ordinals below mirror the enumerations
//! `ffi.c` kept file-local. `test/ffi_layout.zig` pins them, by writing the
//! numbers out rather than by importing these declarations.

const std = @import("std");
const abi = @import("abi");

/// `JanetFFIPrimType` in `src/core/ffi.c`.
const PrimType = enum(i32) {
    void = 0,
    bool = 1,
    ptr = 2,
    string = 3,
    float = 4,
    double = 5,
    int8 = 6,
    uint8 = 7,
    int16 = 8,
    uint16 = 9,
    int32 = 10,
    uint32 = 11,
    int64 = 12,
    uint64 = 13,
    @"struct" = 14,
};

/// `JanetFFICallingConvention` in `src/core/ffi.c`.
const CallingConvention = enum(i32) {
    none = 0,
    sysv64 = 1,
    win64 = 2,
    aapcs64 = 3,
};

/// `JANET_64` in `src/include/janet.h`, which selects the width of the `size`
/// and `ssize` machine types. It is read from the header rather than inferred
/// from the target so that the two cannot disagree.
const is_64_bit = @hasDecl(abi.c, "JANET_64");

const NamedPrim = struct { []const u8, PrimType };

/// Every machine-type name `ffi.c` accepts, in its order: the primary names,
/// the word-size-dependent pair, then the aliases. `struct` has no name of its
/// own — a struct type is written as a tuple rather than a keyword.
const prim_names = [_]NamedPrim{
    .{ "void", .void },
    .{ "bool", .bool },
    .{ "ptr", .ptr },
    .{ "pointer", .ptr },
    .{ "string", .string },
    .{ "float", .float },
    .{ "double", .double },
    .{ "int8", .int8 },
    .{ "uint8", .uint8 },
    .{ "int16", .int16 },
    .{ "uint16", .uint16 },
    .{ "int32", .int32 },
    .{ "uint32", .uint32 },
    .{ "int64", .int64 },
    .{ "uint64", .uint64 },
    .{ "size", if (is_64_bit) .uint64 else .uint32 },
    .{ "ssize", if (is_64_bit) .int64 else .int32 },
    .{ "r32", .float },
    .{ "r64", .double },
    .{ "s8", .int8 },
    .{ "u8", .uint8 },
    .{ "s16", .int16 },
    .{ "u16", .uint16 },
    .{ "s32", .int32 },
    .{ "u32", .uint32 },
    .{ "s64", .int64 },
    .{ "u64", .uint64 },
    .{ "char", .int8 },
    .{ "short", .int16 },
    .{ "int", .int32 },
    .{ "long", .int64 },
    .{ "byte", .uint8 },
    .{ "uchar", .uint8 },
    .{ "ushort", .uint16 },
    .{ "uint", .uint32 },
    .{ "ulong", .uint64 },
};

const NamedCc = struct { []const u8, CallingConvention };

/// The calling conventions that have a name of their own. `default` is not
/// here: it resolves to whichever convention the build enables, which is a
/// property of the target rather than of the name, so `ffi_types.zig` maps it.
const cc_names = [_]NamedCc{
    .{ "none", .none },
    .{ "sysv64", .sysv64 },
    .{ "win64", .win64 },
    .{ "aapcs64", .aapcs64 },
};

// ---------------------------------------------------------------------------
// Name decoding
// ---------------------------------------------------------------------------

/// Report the machine type a keyword names, or -1 when it names none.
///
/// The keyword arrives as bytes rather than as a C string because a Janet
/// keyword is length-prefixed and may contain a zero byte; that is also what
/// `janet_cstrcmp` compared, so equality here means exactly what equality meant
/// before.
///
/// Every name is decoded on every target, including the two whose meaning
/// depends on the word size. The ordinal is answered rather than a `PrimType`,
/// because the caller's own enumeration is the one the rest of the FFI speaks
/// and mapping into it is `ffi_types.zig`'s business — an unknown name is a
/// Janet error there, and it has the value to name.
pub fn decodePrim(name: []const u8) i32 {
    for (prim_names) |entry| {
        if (std.mem.eql(u8, name, entry[0])) return @intFromEnum(entry[1]);
    }
    return -1;
}

/// Report the calling convention a keyword names, or -1 when it names none.
///
/// Unlike the C original, this decodes all four names everywhere. Whether a
/// convention is *enabled* is a property of the target, and `ffi_types.zig`
/// still decides that; folding the two questions together is what left the
/// conventions a build does not enable with no coverage at all.
pub fn decodeCc(name: []const u8) i32 {
    for (cc_names) |entry| {
        if (std.mem.eql(u8, name, entry[0])) return @intFromEnum(entry[1]);
    }
    return -1;
}

// ---------------------------------------------------------------------------
// Type extent
// ---------------------------------------------------------------------------

/// The number of bytes an array of `array_count` elements of `base_size`
/// occupies, where a negative count means the type is not an array.
///
/// The multiplication wraps explicitly. A count comes from `janet_getnat` and a
/// base size from a completed layout, so Janet's own limits keep the product
/// well inside the range; but C leaves an overflow of `size_t` defined and Zig
/// would trap, so the port commits to the C result.
pub fn typeExtent(base_size: usize, array_count: i32) usize {
    const count: usize = if (array_count < 0) 1 else @intCast(array_count);
    return base_size *% count;
}

// ---------------------------------------------------------------------------
// Struct layout
// ---------------------------------------------------------------------------

/// The running state of a struct layout: the bytes placed so far, the strictest
/// alignment any field has demanded, and whether every field has so far landed
/// on its natural boundary.
///
/// `ffi_types.zig` holds the `Struct` this ends up in, and every field offset
/// reported here is written there. The three counters are `u32` because that is
/// what a `Struct` stores them in, and the arithmetic truncates to that width
/// exactly where the C implementation's casts did.
pub const Layout = extern struct {
    size: u32,
    alignment: u32,
    is_aligned: u32,

    /// `janet_ffi_layout_init`.
    pub fn init() Layout {
        return .{ .size = 0, .alignment = 1, .is_aligned = 1 };
    }

    /// Place one field and report the offset it takes.
    ///
    /// `el_align` must not be zero; `ffi_types.zig` rejects a zero-aligned
    /// field with a panic immediately before calling, because that panic names
    /// the offending Janet value.
    ///
    /// A packed field is laid down where the cursor stands and contributes
    /// nothing to the struct's alignment, but a packed field that happens to
    /// land off its natural boundary clears `is_aligned` — that flag records
    /// whether the layout could have been produced without packing, which is
    /// what the callback path later relies on.
    ///
    /// `packed_field` is a `bool` rather than the C original's `int`. The one
    /// caller builds it from two keyword tests, and the flag deciding padding
    /// is the sort of argument a bare `0` at a call site says nothing about.
    pub fn place(self: *Layout, el_size: usize, el_align: usize, packed_field: bool) usize {
        if (packed_field) {
            if (@as(usize, self.size) % el_align != 0) self.is_aligned = 0;
            const offset = self.size;
            self.size = self.size +% @as(u32, @truncate(el_size));
            return offset;
        }
        if (el_align > self.alignment) self.alignment = @truncate(el_align);
        const aligned = ((@as(usize, self.size) +% el_align -% 1) / el_align) *% el_align;
        const offset: u32 = @truncate(aligned);
        self.size = @truncate(el_size +% offset);
        return offset;
    }

    /// Round the total up to the struct's own alignment, which is what makes an
    /// array of the struct place every element correctly.
    pub fn finish(self: *Layout) void {
        self.size = self.size +% (self.alignment -% 1);
        self.size = self.size / self.alignment;
        self.size = self.size *% self.alignment;
    }
};

test "every machine type name decodes and unknown names are reported" {
    try std.testing.expectEqual(@as(i32, 0), decodePrim("void"));
    try std.testing.expectEqual(@as(i32, 2), decodePrim("pointer"));
    try std.testing.expectEqual(@as(i32, -1), decodePrim("nonesuch"));
    try std.testing.expectEqual(@as(i32, -1), decodePrim("voidx"));
    try std.testing.expectEqual(@as(i32, -1), decodePrim("voi"));
}

test "every calling convention decodes regardless of target" {
    try std.testing.expectEqual(@as(i32, 1), decodeCc("sysv64"));
    try std.testing.expectEqual(@as(i32, 2), decodeCc("win64"));
    try std.testing.expectEqual(@as(i32, 3), decodeCc("aapcs64"));
    try std.testing.expectEqual(@as(i32, -1), decodeCc("default"));
}

test "layout pads to alignment and rounds the total" {
    var layout = Layout.init();
    try std.testing.expectEqual(@as(usize, 0), layout.place(1, 1, false));
    try std.testing.expectEqual(@as(usize, 8), layout.place(8, 8, false));
    layout.finish();
    try std.testing.expectEqual(@as(u32, 16), layout.size);
    try std.testing.expectEqual(@as(u32, 8), layout.alignment);
    try std.testing.expectEqual(@as(u32, 1), layout.is_aligned);
}

test "a packed field off its boundary clears the aligned flag" {
    var layout = Layout.init();
    try std.testing.expectEqual(@as(usize, 0), layout.place(1, 1, true));
    try std.testing.expectEqual(@as(usize, 1), layout.place(8, 8, true));
    layout.finish();
    try std.testing.expectEqual(@as(u32, 9), layout.size);
    try std.testing.expectEqual(@as(u32, 1), layout.alignment);
    try std.testing.expectEqual(@as(u32, 0), layout.is_aligned);
}

File: test/ffi_layout.c
/* Behavioral contract for the FFI type system's portable kernels, run against
 * whichever implementation the build selected (`-Dffi-layout=c` or the Zig
 * default).
 *
 * The name tables are pinned name by name, including every alias, because a
 * table is exactly the kind of thing a port drops one entry from. The layout
 * machine is checked two ways: against fixed vectors, and against the offsets
 * the host C compiler itself assigns to equivalent structures. The second is
 * the stronger check — it says the machine reproduces the platform's ABI
 * rather than merely reproducing whatever the previous implementation did.
 *
 * All four calling-convention names are asserted here even though a build
 * enables at most one of them. Deciding which are enabled stays in `ffi.c`;
 * decoding a name does not depend on the target, and asserting that on every
 * target is the coverage the arch-gated C original could never have.
 */

#include <assert.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <janet.h>

/* Declared rather than included from src/core/ffi.c, so the contract depends
 * only on the internal ABI it exercises. A compile-time assertion in that file
 * pins the ordinals below to the enumerations they mirror. */
typedef struct {
    uint32_t size;
    uint32_t alignment;
    uint32_t is_aligned;
} JanetFFILayout;

enum {
    PRIM_VOID = 0,
    PRIM_BOOL = 1,
    PRIM_PTR = 2,
    PRIM_STRING = 3,
    PRIM_FLOAT = 4,
    PRIM_DOUBLE = 5,
    PRIM_INT8 = 6,
    PRIM_UINT8 = 7,
    PRIM_INT16 = 8,
    PRIM_UINT16 = 9,
    PRIM_INT32 = 10,
    PRIM_UINT32 = 11,
    PRIM_INT64 = 12,
    PRIM_UINT64 = 13,
    PRIM_STRUCT = 14
};

enum {
    CC_NONE = 0,
    CC_SYSV_64 = 1,
    CC_WIN_64 = 2,
    CC_AAPCS64 = 3
};

int32_t janet_ffi_decode_prim(const uint8_t *name, int32_t len);
int32_t janet_ffi_decode_cc(const uint8_t *name, int32_t len);
size_t janet_ffi_type_extent(size_t base_size, int32_t array_count);
void janet_ffi_layout_init(JanetFFILayout *layout);
size_t janet_ffi_layout_place(JanetFFILayout *layout, size_t el_size, size_t el_align, int packed_field);
void janet_ffi_layout_finish(JanetFFILayout *layout);

static int32_t prim(const char *name) {
    return janet_ffi_decode_prim((const uint8_t *) name, (int32_t) strlen(name));
}

static int32_t cc(const char *name) {
    return janet_ffi_decode_cc((const uint8_t *) name, (int32_t) strlen(name));
}

/* ------------------------------------------------------------ name tables */

static void test_primary_machine_types(void) {
    assert(prim("void") == PRIM_VOID);
    assert(prim("bool") == PRIM_BOOL);
    assert(prim("ptr") == PRIM_PTR);
    assert(prim("pointer") == PRIM_PTR);
    assert(prim("string") == PRIM_STRING);
    assert(prim("float") == PRIM_FLOAT);
    assert(prim("double") == PRIM_DOUBLE);
    assert(prim("int8") == PRIM_INT8);
    assert(prim("uint8") == PRIM_UINT8);
    assert(prim("int16") == PRIM_INT16);
    assert(prim("uint16") == PRIM_UINT16);
    assert(prim("int32") == PRIM_INT32);
    assert(prim("uint32") == PRIM_UINT32);
    assert(prim("int64") == PRIM_INT64);
    assert(prim("uint64") == PRIM_UINT64);
}

static void test_machine_type_aliases(void) {
    assert(prim("r32") == PRIM_FLOAT);
    assert(prim("r64") == PRIM_DOUBLE);
    assert(prim("s8") == PRIM_INT8);
    assert(prim("u8") == PRIM_UINT8);
    assert(prim("s16") == PRIM_INT16);
    assert(prim("u16") == PRIM_UINT16);
    assert(prim("s32") == PRIM_INT32);
    assert(prim("u32") == PRIM_UINT32);
    assert(prim("s64") == PRIM_INT64);
    assert(prim("u64") == PRIM_UINT64);
    assert(prim("char") == PRIM_INT8);
    assert(prim("short") == PRIM_INT16);
    assert(prim("int") == PRIM_INT32);
    assert(prim("long") == PRIM_INT64);
    assert(prim("byte") == PRIM_UINT8);
    assert(prim("uchar") == PRIM_UINT8);
    assert(prim("ushort") == PRIM_UINT16);
    assert(prim("uint") == PRIM_UINT32);
    assert(prim("ulong") == PRIM_UINT64);
}

/* The only two names whose meaning depends on the word size. */
static void test_word_sized_machine_types(void) {
#ifdef JANET_64
    assert(prim("size") == PRIM_UINT64);
    assert(prim("ssize") == PRIM_INT64);
#else
    assert(prim("size") == PRIM_UINT32);
    assert(prim("ssize") == PRIM_INT32);
#endif
}

static void test_unknown_machine_types(void) {
    assert(prim("nonesuch") == -1);
    assert(prim("") == -1);
    assert(prim("struct") == -1);       /* written as a tuple, never as a name */
    assert(prim("VOID") == -1);         /* the table is case sensitive */

    /* A prefix of a name and a name with something appended are both unknown:
     * the comparison is on the whole length, not on a leading run. */
    assert(janet_ffi_decode_prim((const uint8_t *) "void", 3) == -1);
    assert(janet_ffi_decode_prim((const uint8_t *) "voidx", 5) == -1);

    /* A keyword may contain a zero byte, so the length is what delimits the
     * name rather than a terminator. */
    assert(janet_ffi_decode_prim((const uint8_t *) "vo\0d", 4) == -1);
    assert(janet_ffi_decode_prim((const uint8_t *) "void\0x", 6) == -1);
}

static void test_calling_conventions(void) {
    /* Every convention decodes on every target, including the three that this
     * build cannot call through. */
    assert(cc("none") == CC_NONE);
    assert(cc("sysv64") == CC_SYSV_64);
    assert(cc("win64") == CC_WIN_64);
    assert(cc("aapcs64") == CC_AAPCS64);

    /* `default` resolves to whichever convention the build enables, which is a
     * property of the target; `ffi.c` maps it before reaching the table. */
    assert(cc("default") == -1);

    assert(cc("nonesuch") == -1);
    assert(cc("") == -1);
    assert(janet_ffi_decode_cc((const uint8_t *) "win64", 4) == -1);
}

/* ----------------------------------------------------------- type extents */

static void test_type_extent(void) {
    /* A negative count means the type is not an array at all. */
    assert(janet_ffi_type_extent(8, -1) == 8);
    assert(janet_ffi_type_extent(0, -1) == 0);
    assert(janet_ffi_type_extent(3, -7) == 3);

    /* `@[type]` with no count decodes to a count of zero, which is a real zero
     * rather than a missing one. */
    assert(janet_ffi_type_extent(8, 0) == 0);

    assert(janet_ffi_type_extent(1, 17) == 17);
    assert(janet_ffi_type_extent(4, 3) == 12);
    assert(janet_ffi_type_extent(8, 1024) == 8192);
}

/* ------------------------------------------------------- layout machinery */

static void test_layout_of_nothing(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(layout.size == 0);
    assert(layout.alignment == 1);
    assert(layout.is_aligned == 1);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 0);
    assert(layout.alignment == 1);
    assert(layout.is_aligned == 1);
}

static void test_layout_pads_between_fields(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 1, 1, 0) == 0);
    /* Seven bytes of padding ahead of the eight-byte field. */
    assert(janet_ffi_layout_place(&layout, 8, 8, 0) == 8);
    assert(janet_ffi_layout_place(&layout, 2, 2, 0) == 16);
    janet_ffi_layout_finish(&layout);
    /* 18 bytes rounded up to the struct's own eight-byte alignment. */
    assert(layout.size == 24);
    assert(layout.alignment == 8);
    assert(layout.is_aligned == 1);
}

static void test_layout_takes_the_strictest_alignment(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 2, 2, 0) == 0);
    assert(janet_ffi_layout_place(&layout, 4, 4, 0) == 4);
    assert(janet_ffi_layout_place(&layout, 1, 1, 0) == 8);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 12);
    assert(layout.alignment == 4);
}

static void test_layout_of_a_single_field(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 4, 4, 0) == 0);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 4);
    assert(layout.alignment == 4);
}

/* An array member contributes its whole extent but only its element's
 * alignment, which is what `janet_ffi_type_extent` and `type_align` produce
 * together. */
static void test_layout_of_an_array_member(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 1, 1, 0) == 0);
    assert(janet_ffi_layout_place(&layout, janet_ffi_type_extent(4, 3), 4, 0) == 4);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 16);
    assert(layout.alignment == 4);
}

static void test_packed_fields_leave_no_padding(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 1, 1, 1) == 0);
    assert(janet_ffi_layout_place(&layout, 8, 8, 1) == 1);
    assert(janet_ffi_layout_place(&layout, 2, 2, 1) == 9);
    janet_ffi_layout_finish(&layout);
    /* Nothing was padded and nothing raised the struct's alignment, so the
     * total is the plain sum. */
    assert(layout.size == 11);
    assert(layout.alignment == 1);
    /* Two of the three landed off their natural boundary. */
    assert(layout.is_aligned == 0);
}

static void test_packed_fields_can_still_be_aligned(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 4, 4, 1) == 0);
    assert(janet_ffi_layout_place(&layout, 4, 4, 1) == 4);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 8);
    /* A packed field contributes nothing to the struct's alignment even when
     * it happens to sit on its own boundary. */
    assert(layout.alignment == 1);
    assert(layout.is_aligned == 1);
}

/* `:pack` packs one member and `:pack-all` packs the rest, so a layout can mix
 * the two kinds of placement. */
static void test_layout_mixes_packed_and_aligned_fields(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 1, 1, 0) == 0);
    assert(janet_ffi_layout_place(&layout, 4, 4, 1) == 1);
    assert(janet_ffi_layout_place(&layout, 8, 8, 0) == 8);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 16);
    assert(layout.alignment == 8);
    assert(layout.is_aligned == 0);
}

/* The rounding at the end is what makes an array of the struct place every
 * element on the alignment its fields demand. */
static void test_layout_rounds_the_total_up(void) {
    JanetFFILayout layout;
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, 8, 8, 0) == 0);
    assert(janet_ffi_layout_place(&layout, 1, 1, 0) == 8);
    assert(layout.size == 9);
    janet_ffi_layout_finish(&layout);
    assert(layout.size == 16);
    assert(layout.size % layout.alignment == 0);
}

/* --------------------------------------------- agreement with the host ABI */

struct host_char_double {
    char a;
    double b;
};

struct host_mixed {
    char a;
    int32_t b;
    char c;
    double d;
    int16_t e;
};

struct host_nested {
    int16_t a;
    struct host_char_double b;
    char c;
};

/* Lay out the equivalent of a C structure and compare every offset and the
 * total against what the host compiler assigned. This is what says the machine
 * reproduces the platform ABI rather than only its own past output. */
static void test_layout_matches_the_host_compiler(void) {
    JanetFFILayout layout;

    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, sizeof(char), sizeof(char), 0)
           == offsetof(struct host_char_double, a));
    assert(janet_ffi_layout_place(&layout, sizeof(double), sizeof(double), 0)
           == offsetof(struct host_char_double, b));
    janet_ffi_layout_finish(&layout);
    assert(layout.size == sizeof(struct host_char_double));
    assert(layout.alignment == sizeof(double));

    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, sizeof(char), sizeof(char), 0)
           == offsetof(struct host_mixed, a));
    assert(janet_ffi_layout_place(&layout, sizeof(int32_t), sizeof(int32_t), 0)
           == offsetof(struct host_mixed, b));
    assert(janet_ffi_layout_place(&layout, sizeof(char), sizeof(char), 0)
           == offsetof(struct host_mixed, c));
    assert(janet_ffi_layout_place(&layout, sizeof(double), sizeof(double), 0)
           == offsetof(struct host_mixed, d));
    assert(janet_ffi_layout_place(&layout, sizeof(int16_t), sizeof(int16_t), 0)
           == offsetof(struct host_mixed, e));
    janet_ffi_layout_finish(&layout);
    assert(layout.size == sizeof(struct host_mixed));

    /* A nested structure enters as its own size and alignment, which is how
     * `type_size` and `type_align` present one. */
    janet_ffi_layout_init(&layout);
    assert(janet_ffi_layout_place(&layout, sizeof(int16_t), sizeof(int16_t), 0)
           == offsetof(struct host_nested, a));
    assert(janet_ffi_layout_place(&layout, sizeof(struct host_char_double),
                                  sizeof(double), 0)
           == offsetof(struct host_nested, b));
    assert(janet_ffi_layout_place(&layout, sizeof(char), sizeof(char), 0)
           == offsetof(struct host_nested, c));
    janet_ffi_layout_finish(&layout);
    assert(layout.size == sizeof(struct host_nested));
}

/* ------------------------------------------------------------- invariants */

/* Sweep every combination of a few sizes and alignments and check the rules
 * that must hold whatever the inputs were, rather than a stored expectation
 * for each one. */
static void test_layout_invariants_over_a_sweep(void) {
    static const size_t alignments[] = {1, 2, 4, 8, 16};
    static const size_t sizes[] = {1, 2, 3, 4, 7, 8, 12, 16, 31};

    for (size_t ai = 0; ai < sizeof(alignments) / sizeof(alignments[0]); ai++) {
        for (size_t si = 0; si < sizeof(sizes) / sizeof(sizes[0]); si++) {
            for (size_t bi = 0; bi < sizeof(alignments) / sizeof(alignments[0]); bi++) {
                for (size_t ti = 0; ti < sizeof(sizes) / sizeof(sizes[0]); ti++) {
                    JanetFFILayout layout;
                    janet_ffi_layout_init(&layout);

                    size_t first = janet_ffi_layout_place(&layout, sizes[si], alignments[ai], 0);
                    assert(first == 0);
                    assert(layout.size == sizes[si]);

                    size_t second = janet_ffi_layout_place(&layout, sizes[ti], alignments[bi], 0);
                    /* Each field starts on its own boundary, never before the
                     * end of the field ahead of it, and never further past it
                     * than that boundary requires. */
                    assert(second % alignments[bi] == 0);
                    assert(second >= sizes[si]);
                    assert(second - sizes[si] < alignments[bi]);
                    assert(layout.size == second + sizes[ti]);

                    uint32_t before_rounding = layout.size;
                    janet_ffi_layout_finish(&layout);

                    /* The struct takes the strictest alignment any field asked
                     * for, its total is a whole number of those, and rounding
                     * never loses a byte or adds a needless one. */
                    size_t expected_align = alignments[ai] > alignments[bi]
                                            ? alignments[ai] : alignments[bi];
                    assert(layout.alignment == expected_align);
                    assert(layout.size % layout.alignment == 0);
                    assert(layout.size >= before_rounding);
                    assert(layout.size - before_rounding < layout.alignment);
                    /* Nothing was packed, so the layout is a natural one. */
                    assert(layout.is_aligned == 1);
                }
            }
        }
    }
}

/* A packed sweep: no padding anywhere, no contribution to the alignment, and
 * the aligned flag reporting exactly whether every field happened to land on
 * its own boundary. */
static void test_packed_layout_invariants_over_a_sweep(void) {
    static const size_t alignments[] = {1, 2, 4, 8, 16};
    static const size_t sizes[] = {1, 3, 4, 5, 8, 13};

    for (size_t ai = 0; ai < sizeof(alignments) / sizeof(alignments[0]); ai++) {
        for (size_t si = 0; si < sizeof(sizes) / sizeof(sizes[0]); si++) {
            for (size_t bi = 0; bi < sizeof(alignments) / sizeof(alignments[0]); bi++) {
                for (size_t ti = 0; ti < sizeof(sizes) / sizeof(sizes[0]); ti++) {
                    JanetFFILayout layout;
                    janet_ffi_layout_init(&layout);

                    assert(janet_ffi_layout_place(&layout, sizes[si], alignments[ai], 1) == 0);
                    size_t second = janet_ffi_layout_place(&layout, sizes[ti], alignments[bi], 1);
                    assert(second == sizes[si]);

                    janet_ffi_layout_finish(&layout);
                    assert(layout.alignment == 1);
                    assert(layout.size == sizes[si] + sizes[ti]);

                    /* The first field sits at zero and so is always natural;
                     * the second is natural exactly when the first field's
                     * size is a multiple of its alignment. */
                    int both_natural = (0 == sizes[si] % alignments[bi]);
                    assert(layout.is_aligned == (uint32_t)(both_natural ? 1 : 0));
                }
            }
        }
    }
}

void ffi_layout_contract(void) {
    test_primary_machine_types();
    test_machine_type_aliases();
    test_word_sized_machine_types();
    test_unknown_machine_types();
    test_calling_conventions();

    test_type_extent();

    test_layout_of_nothing();
    test_layout_pads_between_fields();
    test_layout_takes_the_strictest_alignment();
    test_layout_of_a_single_field();
    test_layout_of_an_array_member();
    test_packed_fields_leave_no_padding();
    test_packed_fields_can_still_be_aligned();
    test_layout_mixes_packed_and_aligned_fields();
    test_layout_rounds_the_total_up();

    test_layout_matches_the_host_compiler();
    test_layout_invariants_over_a_sweep();
    test_packed_layout_invariants_over_a_sweep();

    printf("ffi_layout: all tests passed\n");
}
